Fix trim analyzer warning for inferred type arguments

sbomer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/86032.

This replaces the syntax-based logic with:
- ISymbol-based logic that looks at generic instantiations in parameter types, return types, and base/interface types
- IOperation-based logic that looks at invocations

The syntax-based logic used to also warn for the types of local variables, but `SymbolKind.Local` isn't supported for `RegisterSymbolAction`. Since this falls into the category of warnings that isn't strictly necessary (based on nativeaot logic which only warns on generic instantiations that can actually lead to code execution), I just left out local variables.

I'm slightly concerned that this could be missing some cases, but at least it works on our existing testcases.

There are two slight differences in warning locations, which I don't expect to cause issues for warning suppressions:
- For invocations, the warning location used to be the generic method name, without the argument list. Using the IOperation-based approach, we use the invocation's location that includes the argument list.
- For properties that use arrow syntax (`RequireMethods<TFields> Property => null`), the ISymbol-based approach warns on the underlying getter method, whose location seems to be the body of the accessor. We used to warn on the return type specifically, but I couldn't find a better way to get the location.

sbomer committed:
Fix trim analyzer warning for inferred type arguments

sbomer committed:
Merge remote-tracking branch 'origin/main' into fixInferredGeneric

sbomer committed:
Add ExpectedWarning for analyzer in new testcase

vitek-karas commented on src/tools/illink/src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs:
```diff
@@ -76,17 +77,79 @@ public override void Initialize (AnalysisContext context)
-					ProcessGenericParameters (context);
-				}, SyntaxKind.GenericName);
+				// Examine generic instantiations in base types and interface list
+				context.RegisterSymbolAction (context => {
+					var type = (INamedTypeSymbol) context.Symbol;
+					if (type.BaseType is INamedTypeSymbol baseType) {
```
What if I put RUC on the type - I don't remember if we decided to still warn on base/interface violation or not.

vitek-karas replied:
Thanks.

vitek-karas commented on src/tools/illink/src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs:
```diff
@@ -76,17 +77,79 @@ public override void Initialize (AnalysisContext context)
+
+					var invocation = (IInvocationOperation) context.Operation;
+					var methodSymbol = invocation.TargetMethod;
+					foreach (var diagnostic in ProcessGenericParameters (methodSymbol, invocation.Syntax.GetLocation()))
+						context.ReportDiagnostic (diagnostic);
```
Just curious - is there a reason to use this pattern of returning a list of diagnostics and iterating over it just to call report. Why not call report inside the method? It seems there's lot of unnecessary code duplication this way.

vitek-karas replied:
Now I remember hitting this myself - it's a shame the context doesn't implement an interface with the parts which are common across all contexts (like `ReportDiagnostics`).

vitek-karas commented on src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/GenericParameterDataFlow.cs:
```diff
@@ -432,7 +434,7 @@ public void PublicFieldsMethodLocalVariable ()
 			}
 
-			[ExpectedWarning ("IL2091", nameof (TypeRequiresPublicMethods<TOuter>), ProducedBy = Tool.Trimmer | Tool.Analyzer)] // NativeAOT_StorageSpaceType
+			[ExpectedWarning ("IL2091", nameof (TypeRequiresPublicMethods<TOuter>), ProducedBy = Tool.Trimmer)]
```
Could you please add a comment?
Keep the `NativeAOT_StorageSpaceType` and add something about the analyzer (that it doesn't handle local variables or something).

vitek-karas commented on src/tools/illink/test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresAccessedThrough.cs:
```diff
@@ -74,6 +74,57 @@ public static void Test ()
 			}
 		}
 
+		class AccessedThroughGenericParameterAnnotation
```
Is there a reason you duplicated the test from `RequiresViaDataFlow`?

vitek-karas reviewed (approved):
In general looks good.

sbomer committed:
Add comment and test for RUC on type

sbomer committed:
Fix test for illink, PR feedback

- Allow illink to match ExpectedWarning on property against a
  warning produced from getter
- Add back NativeAOT_StorageSpaceType for illink warning
- Add comment about analyzer local variable behavior

sbomer committed:
Add tests and analyzer support for delegate creation
